### PR TITLE
Heightmap physics (with DART) 🏔️

### LINKED ISSUE #237: Support heightmaps
We should add support for creating terrain based on heightmaps. This is also relevant for supporting DEMs. See issue #235.

### PR DESCRIPTION
**Pull request by chapulina**
# 🎉 New feature

Closes #237

## Summary

Add heightmap support to the physics system.

Requires https://github.com/ignitionrobotics/ign-physics/pull/218

There's a resolution issue still to be solved on the physics side. It looks like only red spheres can stay on top of the heightmap :wink: :red_circle: 

heightmap_dart)opt

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests: **TODO**
- [x] Added example and/or tutorial
- [x] Updated documentation (as needed)
- [ ] ~~Updated migration guide (as needed)~~
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

**Comment by chapulina** [comment]
This is ready for review. The heightmap behaves as expected when using the `bullet` collision checker:

heightmap_bullet

I tested the above with these branches:

* https://github.com/ignitionrobotics/ign-physics/compare/main...chapulina/4/heightmap_detector, which includes https://github.com/ignitionrobotics/ign-physics/pull/218 and https://github.com/ignitionrobotics/ign-physics/pull/25
* https://github.com/ignitionrobotics/ign-gazebo/compare/main...chapulina/5/heightmap_detector, which includes this and https://github.com/ignitionrobotics/ign-gazebo/pull/684

File: src/systems/physics/Physics.cc [hunk 1]
```diff
@@ -879,6 +894,54 @@ void PhysicsPrivate::CreatePhysicsEntities(const EntityComponentManager &_ecm)
...
+              this->entityLinkMap.EntityCast<HeightmapFeatureList>(
+                  _parent->Data());
+          if (!linkHeightmapFeature)
+          {
+            static bool informed{false};
+            if (!informed)
+            {
+              igndbg << "Attempting to process heightmap geometries, but the physics"
+                     << " engine doesn't support feature "
+                     << "[AttachHeightmapShapeFeature]. Heightmapes will be ignored."
```
Comment (nkoenig) [inline src/systems/physics/Physics.cc:909]:
```suggestion
                     << "[AttachHeightmapShapeFeature]. Heightmaps will be ignored."
```

**Comment by chapulina** [comment]
Removed from **beta**, this can come in a minor release.

File: examples/worlds/heightmap.sdf [hunk 2]
```diff
@@ -23,7 +26,113 @@
...
+          <anchors target="3D View">
+            <line own="left" target="left"/>
+            <line own="bottom" target="bottom"/>
+          </anchors>
+        </ignition-gui>
+
+        <play_pause>true</play_pause>
+        <step>true</step>
+        <start_paused>true</start_paused>
+
```
Comment (ahcorde) [inline examples/worlds/heightmap.sdf:52]:
```suggestion
```

File: examples/worlds/heightmap.sdf [hunk 3]
```diff
@@ -23,7 +26,113 @@
...
+            <line own="right" target="right"/>
+            <line own="bottom" target="bottom"/>
+          </anchors>
+        </ignition-gui>
+
+        <sim_time>true</sim_time>
+        <real_time>true</real_time>
+        <real_time_factor>true</real_time_factor>
+        <iterations>true</iterations>
+
```
Comment (ahcorde) [inline examples/worlds/heightmap.sdf:76]:
```suggestion
```

File: examples/worlds/heightmap.sdf [hunk 4]
```diff
@@ -42,6 +151,154 @@
...
+          <mass>3.0</mass>
+        </inertial>
+        <collision name="sphere_collision">
+          <geometry>
+            <sphere>
+              <radius>0.5</radius>
+            </sphere>
+          </geometry>
+        </collision>
+
```
Comment (ahcorde) [inline examples/worlds/heightmap.sdf:175]:
```suggestion
```

File: examples/worlds/heightmap.sdf [hunk 5]
```diff
@@ -42,6 +151,154 @@
...
```
Comment (ahcorde) [inline examples/worlds/heightmap.sdf:212]:
```suggestion
```

File: examples/worlds/heightmap.sdf [hunk 6]
```diff
@@ -42,6 +151,154 @@
...
```
Comment (ahcorde) [inline examples/worlds/heightmap.sdf:249]:
```suggestion
```

File: examples/worlds/heightmap.sdf [hunk 7]
```diff
@@ -42,6 +151,154 @@
...
```
Comment (ahcorde) [inline examples/worlds/heightmap.sdf:286]:
```suggestion
```

**Review by ahcorde — CHANGES_REQUESTED** [review]
I can't see any test, do you plan to add some ?

Some minor style improvements

**Comment by vatanaksoytezer** [comment]
Just for logging purposes, I've tested this PR with a source built fortress and had been able to reproduce the example and also use custom heightmaps with it. It seemed to work pretty well for me. The only down side seems to be is being stuck with ogre rather than ogre2, cross referencing https://github.com/ignitionrobotics/ign-rendering/issues/187 for more information.

**Comment by scpeters** [comment]
we can test this branch with the ignition-physics4 prerelease if we merge the following: https://github.com/ignition-tooling/gzdev/pull/33

**Comment by scpeters** [comment]
the Ignition-physics4 prerelease homebrew bottles are built, so I restarted the macOS job

File: src/systems/physics/Physics.cc [hunk 8]
```diff
@@ -1028,6 +1043,54 @@ void PhysicsPrivate::CreatePhysicsEntities(const EntityComponentManager &_ecm)
...
+        {
+          auto linkHeightmapFeature =
+              this->entityLinkMap.EntityCast<HeightmapFeatureList>(
+                  _parent->Data());
+          if (!linkHeightmapFeature)
+          {
+            static bool informed{false};
+            if (!informed)
+            {
+              igndbg << "Attempting to process heightmap geometries, but the physics"
```
Comment (scpeters) [inline src/systems/physics/Physics.cc:1056]: some code check complaints:

~~~
+ make codecheck
5606
  Scanning dependencies of target cpplint
5607
  /github/workspace/src/systems/physics/Physics.cc:1056:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
5608
  /github/workspace/src/systems/physics/Physics.cc:1058:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
5609
  /github/workspace/src/systems/physics/Physics.cc:1073:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
5610
  /github/workspace/src/systems/physics/Physics.cc:1088:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
5611
  /github/workspace/src/systems/physics/Physics.cc:1326:  Small and focused functions are preferred: PhysicsPrivate::CreatePhysicsEntities() has 527 non-comment lines (error triggered by exceeding 500 lines).  [readability/fn_size] [1]
5612
  Total errors found: 5
~~~

Comment (chapulina) [reply src/systems/physics/Physics.cc]: 9b0d908fa7a460ab0f72480d053e5165e7687358

**Review by scpeters — APPROVED** [review]

**Pull request merged**